README
pstd (pasted) - command line pastebin (version 0.0.2)
=====================================================


Quickstart for running a pstd paste site
----------------------------------------
(To be documented in detail later, but as a quick example, this is how
http://paste.pr0.tips/ is run (referring to it as example.com henceforth)

# Optional: authbind setup (assuming it is installed)
$ sudo touch /etc/authbind/byport/80
$ sudo chown $(whoami) /etc/authbind/byport/80
$ sudo chmod 100 /etc/authbind/byport/80
# (authbind's policy is that a user may bind to a the port if they have
# execute permissions on the file /etc/authbind/byport/<theport>)

# Site setup:
$ git clone https://github.com/fstd/pstd.git
$ cd pstd
$ mkdir pastes
# Then depending on whether or not authbind was used, one of the following:
$ ./pstd_server.pl -l 0.0.0.0:8080 -H example.com:8080  #without authbind
$ authbind --deep ./pstd_server.pl -l 0.0.0.0:80 -H example.com  #with authbind

If that worked, http://example.com/ (or http://example.com:8080/) should contain
further information on how to submit pastes


Quirks
------
I didn't want to run a httpd in front of this because what we really need from
HTTP is but a tiny subset of the protocol; hence a httpd (or even a HTTP lib)
would add considerable attack surface.  Therefore pstd_server.pl generates and
parses the few HTTP messages we support itself.  For a web browser looking at
pastes, this shouldn't matter at all, but for submitting pastes (which can't,
by design, be done in a web browser), there's one major consequence:
The only way to submit a paste is to submit it as a POST request *without any
non-identity "Transfer-Encoding" (like, chunked, or multipart w/ boundary)*.
This implies that the POST request has a "Content-Length" header.
This is the reason for why curl(1) cannot be used to submit pastes like on
sprunge; wget(1) (--post-file) has to be used instead.  However, wget(1) comes
with its own problem of being unable to POST standard input (for it needs to
know the data length in advance).
To cope with this problems, a POSIX shell script "pstd.sh" has been written to
facilitate more convenient pasting (including standard input; it's essentially
a wrapper around wget(1) using a temporary file).  This script is automatically
made available as paste "0" and referred to in the (web-visible) manual page.


This and that
-------------
- The paste IDs we hand out start as 2-letter permutations of [A-Za-z0-0]. Once
  that namespace is (more or less) full (as determined by 10 failed attempts to
  generate an unused ID of that length), the ID length will be raised to three,
  and so forth, up to 32.  This provides plenty or namespace to make sure that
  before we run out of IDs, we'll run out of hard disk space (or sunlight...).

- There's a upper limit on how large a single paste may be, defaults to 256 KB

- Requesting the site root will produce a man page with further information on
  how to submit pastes

- To avoid running as root but still be able to bind to a low-numbered port,
  a NAT setup can be used, or on Linux, authbind(1) has been shown to work.

- Rate limiting is controlled by the -r and -R command line options.
  The argument to -R specifies the size of a (sliding) time window to be
  considered (in seconds), while -r gives the maximum number of attempts to
  paste that may happen (per IP) in that time window. Attempting to paste
  while already rate-limited may extend the period the offender has to wait.
  Note that the amount of memory required per pasting IP goes up linearly
  with the argument to -r (but not -R), so if memory is a concern, better
  turn down -R rather than turning up -r to relax the rate limiting.

- Using the -C and -D command line switches, programs for compression and
  decompression of pastes can be specified.  The arguments to -C and -D should
  be strings representing the command-lines (incl. switches, if any) to
  invoke a compressor and a decompressor, respectively, in a way that they
  act as filter (i.e. (de)compress stdin to stdout).
  For example: -C gzip -D gzcat
  DO NOT CHANGE THE USED COMPRESSION ALGORITHM FOR EXISTING PASTEDIRS!
  By default, no compression is done (technically, it's -C cat -D cat).
  Since these programs are shell-ed out, be careful what you specify.
  The script does not care if it's 'rm -rf ~' or a fork bomb.

- If a link to a paste is suffixed with ?<lang>, then syntax hilighting for
  language <lang> is generated, so long as the site has a program that does
  that.  This is controlled by the -y command line switch, the argument of
  which must be a comma separated string of <lang>:<commandline> tokens,
  where <commandline> represents the way to invoke an external filter that,
  when presented with a source file on stdin, produces formatted HTML on stdout.
  E.g. -y 'c:chilight -f html_color,java:javaprettifier -foobar'

Reasons for writing this
------------------------
On 2015-05-15, sprunge.us was unavailable for about 24 hours, and at that time
was used as the primary (and only) paste backend for a number of scripts (most
notably http://github.com/fstd/cwarn).  I considered running the sprunge code
(http://github.com/rupa/sprunge) but decided against it because I'm not so fond
of python and/or depending on google's cloud storage (or any cloud storage for
that matter).  Regular web-based pastebins also didn't fit my use-case for they
make it unnecessarily difficult to submit data without a web browser.


Todo
----
- There is no mechanism for limited-lifespan pastes yet

- In-memory caching of the most frequenly requested pastes?


Author + License
----------------
(C) 2015, Timo Buhrmester

pstd is licensed under a 2-clause BSD license, without any warranty, but with
best wishes and in the hope it will be useful (also see the LICENSE file).


NEWS
2015-05-16 - Started development + 0.0.1 release
2015-05-17 - 0.0.2 release - Pretty solid already


pstd.sh
#!/bin/sh
# Wrapper around wget(1) for convenient pasting to pstd pastebins

# If -x is given as the *first* argument, enable shell tracing
if [ "$1" = "-x" ]; then shift; set -x; fi

prgnam="$(basename "$0")"
version='0.0.2'


Usage() # Print usage statement and exit
{
	printf 'Usage %s [-h] [-s <pastesite>] [<file>]\n' "$prgnam" >&2
	printf '  If <file> is absent or `-`, we paste stdin.\n' >&2
	printf '  The default paste site is `%s`.\n\n' "$site" >&2
	printf 'v%s, 2015, Timo Buhrmester\n' "$version" >&2
	exit 1
}

Bomb() # Complain loudly and exit
{
	printf '%s: ERROR: %s\n' "$prgnam" "$1" >&2
	exit 1
}

# Check if wget is present, abort if not
which wget >/dev/null || Bomb "We need wget"

# The default paste site to use.  When pstd_server.pl arranges for this script
# to be available as paste "0", this variable is automatically rewritten to
# whatever argument to -H was supplied, or else to what hostname(1) said.
site='127.0.0.1:8080'


# We'll need these two tempfiles later and arrange for them to be rm'ed on exit
tmpin="$(mktemp /tmp/paste.XXXXXXXX)"
tmpout="$(mktemp /tmp/paste.XXXXXXXX)"
trap "rm -f '$tmpin' '$tmpout'" EXIT


# Parse command line arguments
while getopts "s:h" i; do
	case "$i" in
	s) site="$OPTARG" ;;
	*) Usage ;;
	esac
done
shift $((OPTIND-1))

# Warn about superfluous argumnents
[ $# -gt 1 ] && printf '%s: Ignoring all but the first argument\n' "$prgnam" >&2

# If a file is given on the command line, paste that. Else, paste standard input.
if [ $# -gt 0 -a "$1" != '-' ]; then
	[ -r "$1" ] || Bomb "Cannot read '$1'"

	in="$1"
else
	cat >"$tmpin"
	in="$tmpin"
fi

# Paste what $in refers to
wget -q -O - --post-file "$in" "http://$site" >"$tmpout" || Bomb "wget failed :O"

# And output the returned link (or error)
cat "$tmpout"

exit 0

#2015, Timo Buhrmester


pstd_server.pl
#!/usr/bin/env perl

use strict;
use warnings;
use v5.10;

use IO::Select;
use IO::Socket::INET;
use Getopt::Std;
use POSIX 'strftime';
use Data::Dumper;
use Digest::MD5 qw(md5_hex);


# Name and version
my $prgnam = $0 =~ s/^.*\///r;
my $version = '0.0.2';

# These are defaults that can be overridden with command-line switches
my $bindaddr = "127.0.0.1"; # -l
my $bindport = 8080; # -l, On Linux, authbind(1) can be used to listen on low ports without having to run as root
my $pastedir = 'pastes'; # -d
my $manpath = 'pstd.1'; # -m
my $cltscript = 'pstd.sh'; # -c, we need to know this because we distribute this as (synthetical) paste "0"
my $verbose = 0; # -v
my $myhost = ''; # If not overridden by -H, we figure it out by running hostname(1)
my $logfile = ''; # Don't log by default
my $maxbuflen = 256*1024; # -s, The largest single paste we'll accept, in bytes
my $compr = 'cat'; # -C, (default is no compression, hence cat)
my $decompr = 'cat'; # -D, (default is no decompression, hence cat)

# Paste-ID alphabet and the shortest length of IDs we'll hand out
my @idalpha = ("A".."Z", "a".."z", "0".."9");
my $minidlen = 2;

# Map client sockets to their respective read buffers (strings)
my %readbuf;

# Map client sockets to the amount of data we expect from them
my %datalen;

# Map md5 digests of recently pasted files to their respective URLs (prefixed
# with a timestamp followed by a blank.  This is to avoid creating unnecessary
# files in the event that some script or person keeps pasting the same thing.
my %recent;

my $hk_interval = 30*60;
my $hk_next = time + $hk_interval;
my $recent_purge_age = 60*60*24;
my $rateinfo_purge_age; # set to 2*$ratetspan later

# Map client IP addresses to arrays containing timestamps of their latest
# $ratesmpl attempts to paste.  Rate-limiting occurs once the difference
# between the first and the last element of such an array is smaller than
# $ratetspan.  Not only successful pastes, but also running into the
# rate-limiter count towards it, so trying to paste when already rate-limited
# might keep the rate-limiting in place for arbitrary timespans.  This is
# considered a feature.
my %rateinfo;
my $ratesmpl = 5;
my $ratetspan = 30;

# getopts
my %opts;

# map from syntax hilighting language names (e.g. "c") to command lines that
# produce HTML on stdout when presented a source file on stdin.  Elements are
# added by the -y command line option
my %hilite;

my $year = '2015';

# Socket and Select objects
my $sck;
my $sel;

my $quit = 0;
my $inforeq;

sub INFO_handler
{
        $inforeq = 1;
}

sub HK_handler
{
        $hk_next = time;
}

sub INT_handler
{
        $quit = 1;
}

sub L;
sub now { return strftime('%Y-%m-%d %H:%M:%S %z', localtime); }
sub W { say STDERR "$prgnam: ".now.": ".($_[0] =~ s/[\r\n]/\$/grm); }
sub E { my $msg = $_[0]; L "ERROR: $msg"; W "ERROR: $msg"; exit 1; }
sub D { W "DBG: $_[0]" if $verbose; }
sub L
{
	return if (!$logfile);

	my $fhnd;
	if (!open $fhnd, '>>', "$logfile") {
		W "Failed to open $logfile for appending: $!";
		return;
	}

	say $fhnd "$prgnam: ".now.": ".($_[0] =~ s/[\r\n]/\$/grm);
	close $fhnd;
}

sub usage
{
	my ($str, $ec) = @_;

	print $str "Usage: $prgnam [-options (see below)]"
	    ." [-d <path>] [-m <path>]\n"
	    ."  -h: Show this usage statement\n"
	    ."  -V: Print version on stdout\n"
	    ."  -v: Be more verbose\n"
	    ."  -l [addr:]port: Listen on port, optionally bind to addr\n"
	    ."  -m path: Path to manual (pstd.1)\n"
	    ."  -d path: Path to paste directory\n"
	    ."  -c path: Path to pstd.sh client-script (becomes a paste\n"
	    ."           referred to by the manpage)\n"
	    ."  -L path: Path to logfile\n"
	    ."  -H FQDN: Our hostname\n"
	    ."  -s size: Maximum paste size in KiB (this isn't exact \n"
	    ."           because it does not account for the HTTP Header\n"
	    ."  -r num:  Rate limit on `num` pastes in `time` (see -R) secs\n"
	    ."  -R time: See -r. Giving 0 (to -r or -R) means no rate-limit\n"
	    ."  -C cmdl: Compression program to use. The argument should\n"
	    ."           be a program (incl. possible switches) that\n"
	    ."           compresses stdin to stdout (e.g. 'gzip -c')\n"
	    ."  -D cmdl: Decompression program to use. The argument should\n"
	    ."           be a program (incl. possible switches) that\n"
	    ."           decompresses stdin to stdout (e.g. 'gzcat')\n"
	    ."           The default is not to compress/decompress.\n"
	    ."           Once decided for a compression format, you cannot\n"
	    ."           turn around and use something else without clearing\n"
	    ."           the paste directory.\n"
	    ."\nv$version, written by Timo Buhrmester, $year\n";
	exit $ec;
}


sub test_compression
{
	my $teststr = "foo bar\n baz";
	my @out = `echo '$teststr' | $compr | $decompr`;

	return join('', @out) eq "$teststr\n"; #echo appends a newline
}


# Generate an unused paste ID
sub gen_id
{
	# We're trying 10 times to obtain an unused 2-letter ID,
	# then 10 times to obtain an unused 3-letter ID, and so forth
	# We also remember when we start needing to use something longer,
	# and avoid searching the apparently full-ish shorter ID-space then.
	foreach my $idlen ($minidlen..32) {
		foreach my $attempt (1..10) {
			my $id = '';
			$id .= $idalpha[rand @idalpha] for 1..$idlen;

			if (! -e "$pastedir/$id") {
				D "Generated ID $id";
				if ($idlen > $minidlen) {
					$minidlen = $idlen;
				}
				return $id;
			}
		}
	}

	return '';
}


# The maps %recent and %rateinfo grow, but never shrink on their own.
# Therefore, we need to occasionally iterate them and purge old stuff
sub housekeep
{
	my $cnt = 0;
	my @drop;
	my $now = time;

	D "Housekeeping";

	foreach my $key (keys %recent) {
		my $time = $recent{$key} =~ s/ .*$//r;

		if ($now - $time > $recent_purge_age) {
			push @drop, $key;
			$cnt++;
		}
	}

	delete $recent{$_} foreach (@drop);

	D "house-kept away $cnt elements from \%recent";

	$cnt = 0;
	@drop = ();
	foreach my $key (keys %rateinfo) {
		my $aref = $rateinfo{$key};
		my $time = $aref->[@$aref - 1];

		if ($now - $time > $rateinfo_purge_age) {
			push @drop, $key;
			$cnt++;
		}
	}

	delete $rateinfo{$_} foreach (@drop);

	D "house-kept away $cnt elements from \%rateinfo";
}


# Register event and return how many seconds the offender needs to wait
# until they are permitted to paste again
sub ratelimit_check
{
	my ($who) = @_;

	return 0 if (!$ratesmpl or !$ratetspan);

	$rateinfo{$who} = [] if not exists $rateinfo{$who};
	my $aref = $rateinfo{$who};

	shift @$aref if @$aref >= $ratesmpl;
	push @$aref, time;

	if (@$aref >= $ratesmpl) {
		my $tdiff = $aref->[@$aref - 1] - $aref->[0];
		if ($tdiff < $ratetspan) { # going too fast...
			return $ratetspan - (time - $aref->[@$aref - 2]);
		}
	}

	return 0; #okay, not rate limited (yet)
}


# Read and return the man page
sub manpage
{
	my $fhnd;
	if (!open $fhnd, '<', $manpath) {
		W "Failed to open $manpath: $!";
		return "ERROR: Manpage not found\n";
	}

	my @lines = <$fhnd>;
	my $man = join '', @lines;
	close $fhnd;

	return $man =~ s/MYHOST/$myhost/rg;
}


# Deal with a client (most likely a web browser) requesting a paste
# Takes socket and ID as parmeters, returns what we're going to
# pass back to the client (i.e. ideally, the actual paste)
sub process_GET
{
	my ($clt, $id, $hilang) = @_;
	my $who = $clt->peerhost();

	if (! -e "$pastedir/$id") {
		W "$who: Requested nonexistant paste $id";
		return "ERROR: No such paste.\n";
	}

	my $add = '';
	if ($hilang and exists $hilite{$hilang}) {
		$add = "| $hilite{$hilang}";
	}

	my @out = `$decompr <'$pastedir/$id' $add`;
	if (${^CHILD_ERROR_NATIVE} != 0) {
		W "$who: Failed to obtain paste '$pastedir/$id'";
		return "ERROR: Failed to load paste\n";
	}

	my $paste = join '', @out;

	D "$who: Got $id";
	L "$who: Got $id";

	return $paste;
}


# Deal with something (most likely paste.sh or wget) submitting
# a paste.  We don't support much HTTP, so this may not work
# with arbitrary clients. (We require a Content-Length header,
# and no transfer-encoding.  wget --post-file http://.. is okay.
sub process_POST
{
	my ($clt) = @_;
	my $who = $clt->peerhost();
	my $whoipp = $who.":".$clt->peerport();

	my $twait = ratelimit_check $who;
	if ($twait) {
		W "$who: Rate limited";
		return "ERROR: Slow down, cowboy.  Try again in $twait seconds (and don't be early)!\n";
	}


	my $id = gen_id;

	if (!$id) {
		W "Out of IDs (what?)";
		return "ERROR: Out of IDs\n";
	}

	my $paste = $readbuf{$whoipp} =~ s/^(.*?)\r\n\r\n//rs;

	if ($paste eq '') {
		W "$who: Empty paste";
		return "ERROR: Empty paste\n";
	}

	my $digest = md5_hex($paste);
	if (exists $recent{$digest}) {
		my $url = $recent{$digest} =~ s/^[^ ]+ //r;
		my $id = $url =~ s/^.*\///r;
		D "$who: Re-pasted $id";
		L "$who: Re-pasted $id";
		return "$url\n";
	}

	`echo '$paste' | $compr >"$pastedir/$id"`;
	if (${^CHILD_ERROR_NATIVE} != 0) {
		W "$who: Failed to paste '$pastedir/$id'";
		return "ERROR: Failed to write error\n";
	}

	D "$who: Pasted $id";
	L "$who: Pasted $id";

	my $url = "http://$myhost/$id";
	$recent{$digest} = time.' '.$url;

	return "$url\n";
}


# This is called once we have a complete header (for a GET)
# or a complete request (for a POST), it dispatches to
# process_GET and process_POST
sub process_dispatch
{
	my ($clt) = @_;
	my $who = $clt->peerhost();
	my $whoipp = $who.":".$clt->peerport();

	my $resp;

	D "$who: Processing '$readbuf{$whoipp}'";

	if ($readbuf{$whoipp} =~ /^POST \//) {
		$resp=process_POST($clt);
	} elsif ($readbuf{$whoipp} =~ /^GET \/([a-zA-Z0-9]+)\b(?:\?([a-z]+))?\b/) {
		$resp=process_GET($clt, $1, $2);
	} elsif ($readbuf{$whoipp} =~ /^GET \/ /) {
		$resp=manpage;
		L "$who: Manpage";
	} else {
		W "$who: Request not understood";
		L "$who: Request not understood: '$readbuf{$whoipp}'";
		$resp = "ERROR: Request not understood\n";
	}

	return $resp;
}


# Read some more data for the given client and call process_dispatch
# on it once we have enough. Bail out if we get too much
# Return 0 to drop the client, 1 to keep going
sub handle_clt
{
	my ($clt) = @_;
	my $who = $clt->peerhost();
	my $whoipp = $who.":".$clt->peerport();

	D "$who: Handling";

	my $data = '';
	$clt->recv($data, 1024); # XXX can this fail?!

	# I suppose empty data means EOF, but not quite sure. XXX
	if ($data eq '') {
		W "$who: Empty read";
		respond($clt, "text/plain", "ERROR: You what?\n");
		return 0;
	}

	# some early sanity check, this assumes the first couple bytes come in in one chunk, though.
	if (!length $readbuf{$whoipp}) {
		if ($data =~ /^POST \//) {
			$datalen{$whoipp} = -1; #don't know yet
		} elsif ($data =~ /^GET \/([a-zA-Z0-9]+(?:\?([a-z]+))?)? HTTP/) {
			$datalen{$whoipp} = 0; #don't care
		} else {
			W "$who: Bad first data chunk '$data'";
			respond($clt, "text/plain", "ERROR: Request not understood\n");
			return 0;
		}
	}

	$readbuf{$whoipp} .= $data;

	my $buflen = length $readbuf{$whoipp};
	if ($buflen > $maxbuflen) {
		W "$who: Too much data ($buflen/$maxbuflen)";
		respond($clt, "text/plain", "ERROR: Too much data\n");
		return 0;
	}

	# $datalen{$who} contains how many bytes we're expecting to receive
	# from $who.  If it is zero, then we read until we have a complete
	# HTTP-header (i.e. till the first \r\n\r\n).  If it is -1, then
	# this is a POST, and we haven't seen the Content-Length header yet.

	if ($datalen{$whoipp} == -1) {
		# POST, see if we have a header yet...
		my $hdr = $readbuf{$whoipp} =~ s/\r\n\r\n.*$//rs;
		if ($hdr) {
			#... and extract the Content-Length; bail if none
			my $match = $hdr =~ /Content-Length: ([0-9]+)/;
			if ($match and !$1) {
				W "$who: No Content-Length in header";
				respond($clt, "text/plain", "ERROR: Need Content-Length Header\n");
				return 0;
			}
			$datalen{$whoipp} = $1 + 4 + length $hdr;
			D "$who: Expecting $datalen{$whoipp} bytes in total";

			# Also complain if we happen to see an unsupported TE
			$match = $hdr =~ /Transfer-Encoding: ([a-zA-Z0-9_-]+)/;
			if ($match and $1) {
				if ($1 ne 'Identity' and $1 ne 'None') {
					W "$who: Bad TE '$1'";
					respond($clt, "text/plain", "ERROR: Bad Transfer-Encoding (use Identity)\n");
					return 0;
				}
			}
		}
	} elsif ($datalen{$whoipp} == 0) {
		# a GET, process_dispatch once we have a complete request
		if ($readbuf{$whoipp} =~ /\r\n\r\n/) {
			my $resp = process_dispatch($clt);

			# hack: deliver manpage as HTML (for links, mainly)
			my $ctype = ($resp =~ /^<html>/) ? "text/html" : "text/plain";
			respond($clt, $ctype, $resp);
			return 0;
		}
	}

	# datalen may have changed at this point (in the above conditional)

	if ($datalen{$whoipp} > 0) {
		if (length $readbuf{$whoipp} == $datalen{$whoipp}) {
			# a POST, we got everything.
			respond($clt, "text/plain", process_dispatch($clt));
			return 0;
		} elsif (length $readbuf{$whoipp} > $datalen{$whoipp}) {
			# a POST, we got more than advertised.
			W "$who: More data than advertised";
			respond($clt, "text/plain", "ERROR: More data than advertised. Nice try?\n");
			return 0;
		}
	}

	return 1;
}


# respond to client with a fake 200 OK and the actual response
sub respond
{
	my ($clt, $ctype, $data) = @_;
	my $who = $clt->peerhost();

	my $len = length $data;
	my $resp = "HTTP/1.1 200 OK\r\n".
	           "Content-Type: $ctype; charset=UTF-8\r\n".
	           "Content-Length: $len\r\n".
	           "Connection: close\r\n\r\n$data";

	if (!$clt->send($resp)) {
		W "$who: send: $!";
	}
}


sub dump_state
{
	say STDERR "============ state dump =============";
	say STDERR "\$^O = '$^O'";
	say STDERR "\$0 = '$0'";
	say STDERR "\$prgnam = '$prgnam'";
	say STDERR "\$version = '$version'";
	say STDERR "\$bindaddr = '$bindaddr'";
	say STDERR "\$bindport = '$bindport'";
	say STDERR "\$pastedir = '$pastedir'";
	say STDERR "\$manpath = '$manpath'";
	say STDERR "\$cltscript = '$cltscript'";
	say STDERR "\$verbose = '$verbose'";
	say STDERR "\$myhost = '$myhost'";
	say STDERR "\$logfile = '$logfile'";
	say STDERR "\$maxbuflen = '$maxbuflen'";
	say STDERR "\$compr = '$compr'";
	say STDERR "\$decompr = '$decompr'";
	say STDERR "\$minidlen = '$minidlen'";
	say STDERR "\$ratesmpl = '$ratesmpl'";
	say STDERR "\$ratetspan = '$ratetspan'";
	say STDERR "\$year = '$year'";
	say STDERR "\$inforeq = '$inforeq'";
	say STDERR "\$quit = '$quit'";
	say STDERR "\$hk_interval = '$hk_interval'";
	say STDERR "\$hk_next = '$hk_next'";
	say STDERR "\$recent_purge_age = '$recent_purge_age'";
	say STDERR "\$rateinfo_purge_age = '$rateinfo_purge_age;'";

	say STDERR "\@idalpha: '".(join '', @idalpha)."'";

	print STDERR Dumper(\%readbuf) =~ s/\$VAR1/%readbuf/r;
	print STDERR Dumper(\%datalen) =~ s/\$VAR1/%datalen/r;
	print STDERR Dumper(\%rateinfo) =~ s/\$VAR1/%rateinfo/r;
	print STDERR Dumper(\%recent) =~ s/\$VAR1/%recent/r;
	print STDERR Dumper(\%opts) =~ s/\$VAR1/%opts/r;

	print STDERR Dumper($sck) =~ s/\$VAR1/\$sck/r;
	print STDERR Dumper($sel) =~ s/\$VAR1/\$sel/r;
	say STDERR "========= End of state dump =========";
}

# -----------------------------------------------------------------------------


# Parse command-line, overriding defaults
usage(\*STDERR, 1) if !getopts("hvVl:d:m:H:c:L:s:r:R:C:D:y:", \%opts);

if (defined $opts{V}) {
	print "$version\n";
	exit 0;
}

usage(\*STDOUT, 0)    if defined $opts{h};
$verbose = 1          if defined $opts{v};
$manpath = $opts{m}   if defined $opts{m};
$pastedir = $opts{d}  if defined $opts{d};
$cltscript = $opts{c} if defined $opts{c};
$myhost = $opts{H}    if defined $opts{H};
$logfile = $opts{L}   if defined $opts{L};
$maxbuflen = $opts{s} if defined $opts{s};
$ratesmpl = $opts{r}  if defined $opts{r};
$ratetspan = $opts{R} if defined $opts{R};
$compr = $opts{C}     if defined $opts{C};
$decompr = $opts{D}   if defined $opts{C};

if (defined $opts{l}) {
	my $tmp = $opts{l};
	if (!($tmp =~ /(?:(^.+):)?([0-9]+)$/)) {
		E 'Bad argument to -l (should be "PORT" or "ADDR:PORT")';
	}
	$bindaddr = $1 ? $1 : "0.0.0.0";
	$bindport = $2;
}

E "Could not read man page '$manpath' (Bad -m? Try -h)" if ! -r $manpath;
E "Could not read client script '$cltscript' (Bad -c? Try -h)" if ! -r $cltscript;
E "Could not access paste directory '$pastedir' (Bad -d? Try -h)" if ! -d $pastedir;
E "Invalid maximum paste size '$maxbuflen' (Bad -s? Try -h)" if !($maxbuflen =~ /^[1-9][0-9]*$/);
$maxbuflen *= 1024 if defined $opts{s}; # Is given in KiB on the command line

E "Invalid number of rate limiting samples '$ratesmpl' (Bad -r? Try -h)" if !($ratesmpl =~ /^[0-9]+$/);
E "Invalid timespan for rate-limiting '$ratetspan' (Bad -R? Try -h)" if !($ratetspan =~ /^[0-9]+$/);
$rateinfo_purge_age = 2*$ratetspan;

L "Logfile created" if $logfile and ! -e $logfile;
E "Cannot write to logfile '$logfile' (Bad -L? Try -h)" if $logfile and ! -w $logfile;

if (!$myhost) {
	my @out = `hostname`;
	$myhost = $out[0];
	chop $myhost;

	E "Failed to figure out hostname, use -H <FQDN>" if !$myhost;
	W "Determined our hostname to be '$myhost' (override with -H)";
}

if (!test_compression) {
	E "Chosen compression/decompression does not work! (see -h)";
}

if (defined $opts{y}) {
	foreach my $hi (split ',', $opts{y}) {
		my $lang = $hi =~ s/:.*$//r;
		my $prog = $hi =~ s/^[^:]*://r;
		my @out = `echo 'int main(void){}' | $prog`;
		if (${^CHILD_ERROR_NATIVE} != 0) {
			W "Ignoring $lang syntax hilighting because 'echo some-C-code | $prog' didn't work";
			next;
		}

		$hilite{$lang} = $prog;
	}
}

D "Will listen on $bindaddr:$bindport; accessible as http://$myhost/";

# Generate paste '0' because it contains the client script we advertise in the man page
`sed "s/^site=.*\$/site='$myhost'/" $cltscript | $compr >$pastedir/0`;
E "Failed to generate paste 0 (the client script)" if (${^CHILD_ERROR_NATIVE} != 0);

W "NOT doing rate-limiting! (Bad -r and/or -R?)" if (!$ratesmpl || !$ratetspan);

$| = 1;

$sck = new IO::Socket::INET (
	Type => SOCK_STREAM,
	Proto => 'tcp',
	Listen => 64,
	Reuse => 1,
	Blocking => 1,
	LocalAddr => $bindaddr,
	LocalPort => $bindport
) or E "Could not create socket $!\n";

$SIG{'INFO'} = 'INFO_handler' if $^O =~ /^(.*bsd)|darwin$/;
$SIG{'USR1'} = 'INFO_handler';
$SIG{'USR2'} = 'HK_handler';
$SIG{'INT'} = 'INT_handler';

$sel = IO::Select->new();
$sel->add($sck);

while (!$quit)
{
	D "Selecting...";
	my @rdbl = $sel->can_read;

	if ($inforeq) {
		dump_state;
		$inforeq = 0;
	}

	last if ($quit);

	if (time >= $hk_next) {
		housekeep
		$hk_next = time + $hk_interval;
	}

	if (!@rdbl) {
		D "Nothing selected";
		next;
	}

	foreach my $s (@rdbl) {
		if ($s == $sck) {
			D "Listener is readable...";
			my $clt = $sck->accept();
			if (!$clt) {
				W "Failed to accept: $!";
				next;
			}
			my $who = $clt->peerhost();
			my $whoipp = $who.":".$clt->peerport();
			D "$who: Connected";

			$sel->add($clt);
			$readbuf{$whoipp} = '';
			delete $datalen{$whoipp};

			next;
		}

		my $who = $s->peerhost();
		my $whoipp = $who.":".$s->peerport();

		D "$who: Readable";

		if (!handle_clt($s, $who)) {
			D "$who: Dropping";
			$sel->remove($s);
			$s->close();
			delete $readbuf{$whoipp};
			delete $datalen{$whoipp};
		}
	}
}

$sck->close();

L "Terminating";
W "Terminating";

#2015, Timo Buhrmester
